### PR TITLE
Remove proxy tests. Add verify callback tests.

### PR DESCRIPTION
**Pull request by ekasper**
The old proxy tests test the implementation of an application proxy
policy callback defined in the test itself, which is not particularly
useful.

It is, however, useful to test cert verify overrides in
general. Therefore, replace these tests with tests for cert verify
callback behaviour.

Also glob the ssl test inputs on the .in files to catch missing
generated files.

### CODE REVIEW

File: test/ssltest_old.c [hunk 1]
```diff
@@ -1074,7 +1067,7 @@ int main(int argc, char *argv[])
     int client_auth = 0;
     int server_auth = 0, i;
     struct app_verify_arg app_verify_arg =
-        { APP_CALLBACK_STRING, 0, 0, NULL, NULL };
+        { APP_CALLBACK_STRING, 0 };
```
Comment (richsalz) [inline test/ssltest_old.c]: does this aggregate initialization work on non-static things?

**Comment by richsalz** [comment]
+1 pending your response to my comment.

**Comment by levitte** [comment]
Wait!  Hold on there...  is there a reason testing of proxy cert policy checking should be removed, more than "not particularly useful"?  Are proxy certs not used at all these days?

**Comment by ekasper** [comment]
@levitte - but I explained *why* it's not useful. The test defines a complicated proxy policy verification callback and then tests that the policy *defined in the test* works as expected. It's not testing any part of the library API, just its own behaviour.

This makes no sense. What we need to test is that app verify callback decisions propagate through correctly. That the app verify callback makes its decisions correctly is something for the application to test.

File: test/ssltest_old.c (hunk 1 above)
Comment (ekasper) [inline test/ssltest_old.c]: Yes. It's not new either, I just removed unused fields.

**Comment by levitte** [comment]
Actually, it does test that a proxy cert chain is checked through, so it does check that the underlying functionlity does its job correctly.  And it does test that it's possible to implement rights validation through the verification callback, which is the only way to check a proxy cert chain and the rights they allow.  Apart from the InheritAll and the Independent policies, all policies are locally defined where they are used.

If you know of some other way to test this particular corner of certification validation, I'm all ears!

File: test/ssltest_old.c (hunk 1 above)
Comment (richsalz) [inline test/ssltest_old.c]: Thanks.  still +1

**Comment by ekasper** [comment]
@levitte: the underlying functionality - the verify callback decision based on the proxy policy and the cert  - is implemented *in the test*. This is tautological - the test is testing itself.

When you abstract away a little: everything that happens between entering and exiting the application callback is application-defined and cannot be tested by us. As far as our tests are concerned, we should treat the callback as a black box that says "yay" or "nay". We must test that the callback is called and its decision is correctly handled - this part is carried over to the new test.

Now the proxy policy callback is simply one such yay/nay box. However, it is a very complicated one - it is better to test with simpler callbacks where the yay/nay behaviour is obviously as advertised. Otherwise you end up testing too many things at once, and can easily end up hiding a handshake bug with a bug in your test.

If you want to demonstrate how to implement a verify callback for proxy certificates, then perhaps this code belongs in a demo?

**Comment by ekasper** [comment]
Specifically, I think the misunderstanding is here: "it does test that it's possible to implement rights validation through the verification callback".

This test does not *test* that it's possible to implement rights validation - it *demonstrates* that it's possible. These are two completely different things. It is possible to do absolutely anything in the yay/nay box, this does not need to be *tested* because what happens in the yay/nay box happens outside the library.

File: test/ssltest_old.c (hunk 1 above)
Comment (ekasper) [inline test/ssltest_old.c]: Thanks - I'll wait until we reach agreement with levitte@.

**Comment by levitte** [comment]
Ah, ok, I do see your point now.  And yes, your idea to move that particular implementation to demos is sane.

Thanks for indulging, I've no further complaints.

File: test/ssltest_old.c (hunk 1 above)
Comment (levitte) [inline test/ssltest_old.c]: Time to act ;-)

**Comment by levitte** [comment]
Ping?

**Comment by ekasper** [comment]
Merged.

**Pull request closed without merging**